### PR TITLE
Try to use attribute protocol for scope.

### PR DESCRIPTION
**Pull request by windelbouwman**
I changed scope to be of PyAttributes type. However, tests fail, and I do not understand why. @cthulahoops I fear it has to do with the attributes resolution system. Could you have a look at why this change produces a failure?

### CODE REVIEW

File: vm/src/frame.rs [hunk 1]
```diff
@@ -577,9 +578,10 @@ impl Frame {
             }
             bytecode::Instruction::StoreLocals => {
                 let locals = self.pop_value();
+                let locals = objdict::py_dict_to_attributes(&locals);
```
Comment (windelbouwman) [inline vm/src/frame.rs]: Okay, the problem is that I create here a copy of the locals dictionary. It should be a reference to the same dict.

**Comment by windelbouwman** [comment]
How about the idea that dict types can have two sorts of payload? A scope payload with `PyAttributes` type, and a fully fledges dict type?

**Comment by OddCoincidence** [comment]
Could you explain this in a little more detail? I'm not super familiar with this part of the code yet.

**Comment by windelbouwman** [comment]
What I want to achieve is to have sort of light weight dictionaries for attributes for classes and instances. I already implemented full dictionaries, and I want to use them as PyObjectPayload for dictionaries. The problem however is that those full dictionaries use `__eq__` and `__hash__` from the objects to properly be able to use any python object that has those methods as a dict key. For instance and class attributes, I would like to keep things simple and use PyAttributes which is a simple String to PyObjectRef hashmap.

I am curious to your opinion on this. Should we have real dicts all the way down? This might be a performance thing, since I think that the rust hashmap has the best performance.

**Comment by OddCoincidence** [comment]
Ah, I see now. If I understand correctly (big if), `mappingproxy` in cpython serves kinda the same purpose for class attributes.

```
>>> class A:
...     pass

>>> a = A()
>>> type(a.__dict__)
<class 'dict'>
>>> type(A.__dict__)
<class 'mappingproxy'>
```

And for instances you can assign non-strings to `__dict__`, so this would break that right?

**Comment by cthulahoops** [comment]
Sorry, I haven't had a chance to do any work on this recently. In short, I think we should use full dictionaries everywhere we need dictionaries.

CPython uses a few optimisations to avoid dictionary lookups on common operations:

* Magic methods are struct members on the class. (```x->tp_call```)
* LOAD_FAST and STORE_FAST turn local variables into array operations.
* ```__slots__``` which I haven't looked into.

My conclusion is use full dictionaries and implement these optimisations.

**Pull request closed without merging**